File: jnet-biz/src/main/java/com/jnet/biz/entity/Image.java
package com.jnet.biz.entity;

import com.baomidou.mybatisplus.annotation.*;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.io.Serializable;
import java.time.LocalDateTime;

/**
 * 图像资产实体
 *
 * @author JNet Team
 * @since 2024-04-16
 */
@Data
@EqualsAndHashCode(callSuper = false)
@TableName("biz_image")
public class Image implements Serializable {

    private static final long serialVersionUID = 1L;

    /**
     * 主键ID（图像ID）
     */
    @TableId(value = "image_id", type = IdType.AUTO)
    private Long imageId;

    /**
     * 所属批次ID
     */
    private Long batchId;

    /**
     * 文件名（存储用）
     */
    private String filename;

    /**
     * 原始文件名（用户上传时的文件名）
     */
    private String originalFilename;

    /**
     * 文件存储路径（兼容旧字段，逐步废弃）
     */
    private String filePath;

    /**
     * 原始文件路径（用户上传的原始文件）
     */
    private String originalFilePath;

    /**
     * 转换后 TIFF 文件路径（仅 JPG/PNG 需要）
     * 如果为空，表示原始文件本身就是 WSI 格式
     */
    private String convertedTiffPath;

    /**
     * 病理报告号
     */
    private String pathologyId;

    /**
     * 患者ID（脱敏处理）
     */
    private String patientId;

    /**
     * 格式 (SVS/NDPI/JPG/PNG)
     */
    private String format;

    /**
     * 生命周期状态 (Raw/Indexed/Processing/Annotated/Verified/Predicted/Archived)
     */
    private String lifecycleStatus;

    /**
     * 标注进度 (0-100)
     */
    private Integer annotationProgress;

    /**
     * 层级
     */
    private Integer levels;

    /**
     * 图像宽度（像素）
     */
    private Integer width;

    /**
     * 图像高度（像素）
     */
    private Integer height;

    /**
     * X轴物理分辨率 (um/px)
     */
    private Double mppX;

    /**
     * Y轴物理分辨率 (um/px)
     */
    private Double mppY;

    /**
     * 放大倍数
     */
    private Integer magnification;

    /**
     * 文件大小（字节）
     */
    private Long fileSize;

    /**
     * 扫描仪详细信息 (JSONB)
     */
    private String scannerInfo;

    /**
     * 扩展元数据 (JSONB)
     */
    private String metadata;

    /**
     * 缩略图 URL
     */
    private String thumbnailUrl;
    
    /**
     * 是否需要转换（JPG/PNG 为 true，WSI 为 false）
     */
    private Boolean requiresConversion;
    
    /**
     * 转换状态: NONE/PENDING/COMPLETED/FAILED
     */
    private String conversionStatus;

    /**
     * 创建人ID
     */
    private Long createBy;

    /**
     * 创建时间
     */
    private LocalDateTime createTime;

    /**
     * 更新人ID
     */
    private Long updateBy;

    /**
     * 更新时间
     */
    private LocalDateTime updateTime;

    /**
     * 删除标志
     */
    @TableLogic
    private Boolean delFlag;
}

File: ui/src/types/image.ts
/**
 * 图像相关类型定义
 */

// 图像格式枚举
export enum ImageFormat {
  SVS = 'SVS',
  TIFF = 'TIFF',
  JPG = 'JPG',
  PNG = 'PNG',
  NDPI = 'NDPI'
}

// 图像实体
export interface Image {
  imageId: number
  filename: string
  originalFilename: string
  batchId: number
  projectId?: number
  pathologyId?: string
  patientId?: string
  format?: string  // SVS/TIFF/JPG/PNG/NDPI
  fileSize?: number
  width?: number
  height?: number
  levels?: number
  lifecycleStatus?: string  // Raw/Indexed/Processing/Annotated/Verified/Predicted/Archived
  annotationProgress?: number // 0-100
  metadata?: string // JSON字符串
  storagePath: string
  thumbnailUrl?: string  // 缩略图URL（数据库字段：thumbnail_url）
  md5?: string
  createBy?: number
  createTime: string
  updateBy?: number
  updateTime: string
}

// 图像元数据VO
export interface ImageMetadataVO {
  imageId: number
  width: number
  height: number
  levels: number
  levelDimensions: Array<{
    level: number
    width: number
    height: number
  }>
  tileWidth: number
  tileHeight: number
  format: string
  magnification?: number
  resolution?: {
    x: number
    y: number
    unit: string
  }
}

// 图像查询DTO
export interface ImageQueryDTO {
  current: number
  size: number
  batchId?: number
  projectId?: number
  pathologyId?: string
  patientId?: string
  lifecycleStatus?: string  // Raw/Indexed/Processing/Annotated/Verified/Predicted/Archived
  format?: string  // SVS/TIFF/JPG/PNG/NDPI
  keyword?: string // 模糊搜索文件名
  startDate?: string
  endDate?: string
}

// 图像状态更新DTO
export interface ImageStatusDTO {
  status: string  // Raw/Indexed/Processing/Annotated/Verified/Predicted/Archived
}

// 标注进度更新DTO
export interface AnnotationProgressDTO {
  imageIds: number[]
  progress: number
}

// 分片上传初始化DTO
export interface ChunkUploadInitDTO {
  fileMd5: string
  filename: string
  fileSize: number
  chunkSize: number
  batchId: number
  pathologyId?: string
  patientId?: string
}

// 分片上传VO
export interface ChunkUploadVO {
  uploadId: string
  fileMd5: string
  chunkSize: number
  totalChunks: number
  uploadedChunks: number[]
  needUpload: boolean // 是否需要上传（秒传检测）
}

// 分片上传DTO
export interface ChunkUploadDTO {
  fileMd5: string
  chunkIndex: number
  chunk: File
}

// Tile查询DTO
export interface TileQueryDTO {
  imageId: number
  level: number
  row: number
  col: number
}

// 批量选择图像DTO
export interface BatchSelectImagesDTO {
  sourceType: 'IMAGE_STORE' | 'EXISTING_BATCH' // 来源类型
  sourceBatchId?: number // 源批次ID（当sourceType为EXISTING_BATCH时）
  imageIds: number[] // 选择的图像ID列表
  targetBatchId: number // 目标批次ID
}
